### PR TITLE
DA-511: RAG with Couchbase and Cohere Notebook

### PR DESCRIPTION
**Pull request by teetangh**

### CODE REVIEW

File: cohere/search.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+from langchain_core.documents import Document
+from langchain_core.globals import set_llm_cache
+from langchain_core.output_parsers import StrOutputParser
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_core.runnables import RunnablePassthrough
+from langchain_couchbase.cache import CouchbaseCache
+from langchain_couchbase.vectorstores import CouchbaseVectorStore
+from langchain_openai import ChatOpenAI, OpenAIEmbeddings
+
+import cohere
```
Comment (lokesh-couchbase) [inline cohere/search.py:21]: langchain-cohere?

**Comment by lokesh-couchbase** [comment]
Other comments are same as in #2

**Comment by lokesh-couchbase** [comment]
Why are you not using Command R+ model for LLM?

File: cohere/search.py (hunk 1 above)
Comment (teetangh) [reply cohere/search.py]: fixed it

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+        "outputId": "dc7c8f75-be6a-4ac7-8094-71a78a26a0ae"
+      },
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "Requirement already satisfied: python-dotenv in /usr/local/lib/python3.10/dist-packages (1.0.1)\n",
+            "Requirement already satisfied: couchbase in /usr/local/lib/python3.10/dist-packages (4.3.0)\n",
+            "Requirement already satisfied: cohere in /usr/local/lib/python3.10/dist-packages (5.6.2)\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:20]: You should only install langchain-cohere

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+            "Requirement already satisfied: aiohttp in /usr/local/lib/python3.10/dist-packages (from datasets) (3.10.1)\n",
+            "Requirement already satisfied: huggingface-hub>=0.21.2 in /usr/local/lib/python3.10/dist-packages (from datasets) (0.23.5)\n",
+            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from datasets) (24.1)\n",
+            "Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.10/dist-packages (from datasets) (6.0.1)\n",
+            "Requirement already satisfied: jsonpatch<2.0,>=1.33 in /usr/local/lib/python3.10/dist-packages (from langchain_core) (1.33)\n",
+            "Requirement already satisfied: langsmith<0.2.0,>=0.1.75 in /usr/local/lib/python3.10/dist-packages (from langchain_core) (0.1.98)\n",
+            "Requirement already satisfied: tenacity!=8.4.0,<9.0.0,>=8.1.0 in /usr/local/lib/python3.10/dist-packages (from langchain_core) (8.5.0)\n",
+            "Requirement already satisfied: openai<2.0.0,>=1.32.0 in /usr/local/lib/python3.10/dist-packages (from langchain_openai) (1.39.0)\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:52]: You should cohere's LLM instetad of OpenAI

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 11,
+      "metadata": {
+        "id": "7eV1X5xILJRC"
+      },
+      "outputs": [],
+      "source": [
+        "def main():\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:391]: Typically in Notebooks, you write the code without functions in smaller sections explaining them as you write the sections.

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "iar2fABrLJjK"
+      },
+      "source": [
+        "# Cache the Results\n",
+        "This function initializes and returns a CouchbaseCache, which is used for caching purposes within the application. The cache helps in reducing redundant computations and improves the performance of operations like search and retrieval."
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:286]: This explanation is quite vague & not accurate.

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:1]: Also, we should create the vector index from scratch.

File: cohere/search.py
Comment (nithishr) [inline cohere/search.py:1]: Why are we replicating the same code in two places (Py & Notebook)?

Comment (teetangh) [reply cohere/search.py:1]: The Python version is modular while the notebook version is non-modular and more focussed to describe the approach and run each each block at a time and see the output

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 3 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb:52]: used cohere's llm

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb:1]: created the vector index from scratch with proper error handling

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 2 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: done

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 5 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: updated with more elaborate explaination

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 4 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: updated

File: cohere/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+numpy
+couchbase
+datasets
+python-dotenv
+langchain-core
+langchain-couchbase
+langchain-openai
```
Comment (nithishr) [inline cohere/requirements.txt:7]: You need to pin the dependencies. 
Also why do you need openai?

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+        "id": "y2H9xphrJLbP"
+      },
+      "execution_count": 10,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# Connect to Couchbase\n",
+        "The script attempts to establish a connection to the Couchbase database using the credentials retrieved from the environment variables. Couchbase is a NoSQL database known for its flexibility, scalability, and support for various data models, including document-based storage. The connection is authenticated using a username and password, and the script waits until the connection is fully established before proceeding.\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:232]: The credentials are not from the environment variables

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 8]
```diff
@@ -0,0 +1,660 @@
...
+      "cell_type": "markdown",
+      "source": [
+        "# Load Index Definition\n",
+        "The search index definition is loaded from a JSON file. This index defines how the data in Couchbase should be indexed for fast search and retrieval. Indexing is critical for optimizing search queries, especially when dealing with large datasets. The JSON file contains details about the index, such as its name, source type, and parameters."
+      ],
+      "metadata": {
+        "id": "j4tYSkkDxS9O"
+      }
+    },
+    {
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:268]: Why is this commented out? Remove it if you don't need

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 9]
```diff
@@ -0,0 +1,660 @@
...
+            "\n",
+            "Financial institution failures played a role in the depression as banks and other financial institutions became insolvent, which led to customers losing their deposits and a lack of lending further deepening the recession. The decline in consumer spending and investment also contributed to the depression as people lost their savings and were unable to spend money on goods and services, leading to decreased economic activity and job losses.\n",
+            "\n",
+            "Other factors included the agricultural crisis of the 1920s, which led to high unemployment rates and poverty, and the failure of President Hoover to address the crisis. Additionally, protective tariffs imposed by the US and other countries further isolated their economies, ultimately worsening the global effects of the depression. \n",
+            "\n",
+            "It is important to note that the causes of the 1929 Great Depression were complex and multi-faceted, and it was a combination of these factors that led to the economic decline.\n",
+            "\n",
+            "Semantic Search Results (completed in 0.53 seconds):\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:644]: I would clean this up before running the notebook to not have these repeated results.

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 10]
```diff
@@ -0,0 +1,660 @@
...
+            "Distance: 0.6208, Text: Why did the world enter a global depression in 1929 ?\n",
+            "Distance: 0.6208, Text: Why did the world enter a global depression in 1929 ?\n",
+            "Distance: 0.6208, Text: Why did the world enter a global depression in 1929 ?\n",
+            "Distance: 0.4913, Text: When was `` the Great Depression '' ?\n",
+            "Distance: 0.4913, Text: When was `` the Great Depression '' ?\n",
+            "Distance: 0.4913, Text: When was `` the Great Depression '' ?\n",
+            "Distance: 0.4913, Text: When was `` the Great Depression '' ?\n",
+            "Distance: 0.3775, Text: What crop failure caused the Irish Famine ?\n",
+            "Distance: 0.3775, Text: What crop failure caused the Irish Famine ?\n"
+          ]
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:655]: Also, can you add an extra cell that showcases the use of the Cache?
Currently, you are caching the response but it is not being used.

File: cohere/search.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+            logging.info(f"Collection '{collection_name}' already exists.")
+        
+        # Wait for the collection to be available
+        max_retries = 3
+        retry_delay = 1
+        for _ in range(max_retries):
+            try:
+                collection = bucket.scope(scope_name).collection(collection_name)
+                break
+            except CollectionNotFoundException:
```
Comment (nithishr) [inline cohere/search.py:88]: Do you get this exception here? It is only creating a reference to the collection here & you would get the exception only when you perform an operation on the collection.

File: cohere/search.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+                time.sleep(retry_delay)
+        else:
+            raise RuntimeError(f"Collection '{collection_name}' not available after {max_retries} retries")
+        
+        # Ensure primary index exists
+        try:
+            cluster.query(f"CREATE PRIMARY INDEX ON `{bucket_name}`.`{scope_name}`.`{collection_name}`").execute()
```
Comment (nithishr) [inline cohere/search.py:95]: You can avoid this error handling by using IF NOT EXISTS clause while creating the index
https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/createprimaryindex.html#if-not-exists

File: cohere/search.py [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+        # Create SearchIndex object
+        search_index = SearchIndex(
+            name=index_definition["name"],
+            source_type=index_definition.get("sourceType", "couchbase"),
+            idx_type=index_definition["type"],
+            source_name=index_definition["sourceName"],
+            params=index_definition["params"],
+            source_params=index_definition.get("sourceParams", {}),
+            plan_params=index_definition.get("planParams", {})
+        )
```
Comment (nithishr) [inline cohere/search.py:144]: Doesn't the `SearchIndex.from_json()` work? 
It should be the recommended approach here as you have the search index json here already.

File: cohere/search.py
Comment (nithishr) [inline cohere/search.py:1]: Can you please add a single line doc string to each function about its functionality

File: cohere/search.py [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+        {"context": vector_store.as_retriever(), "question": RunnablePassthrough()}
+        | prompt
+        | llm
+        | StrOutputParser()
+    )
+    logging.info("Successfully created RAG chain")
+    return chain
+
+
+def main():
```
Comment (nithishr) [inline cohere/search.py:274]: You can remove this function & instead use the script content under the script check. The function does not do anything special here.

**Review by nithishr — CHANGES_REQUESTED** [review]

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 7 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb:2417]: Now i have implemented it from the getpass.getpass method you earlier suggested

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 9 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: cleared the collection now before using it

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 10 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb:3693]: Shown the use case of cache

File: cohere/search.py (hunk 11 above)
Comment (teetangh) [reply cohere/search.py]: removed this exception handling. Had added it as a precautionary measure

File: cohere/search.py (hunk 12 above)
Comment (teetangh) [reply cohere/search.py]: Used the "IF NOT EXISTS" clause now

File: cohere/search.py (hunk 13 above)
Comment (teetangh) [reply cohere/search.py]: Used the `SearchIndex.from_json()` method now

File: cohere/search.py (hunk 14 above)
Comment (teetangh) [reply cohere/search.py]: removed it

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 15]
```diff
@@ -0,0 +1,3514 @@
...
+            "cudf-cu12 24.4.1 requires pyarrow<15.0.0a0,>=14.0.1, but you have pyarrow 17.0.0 which is incompatible.\n",
+            "ibis-framework 8.0.0 requires pyarrow<16,>=2, but you have pyarrow 17.0.0 which is incompatible.\u001b[0m\u001b[31m\n",
+            "\u001b[0mSuccessfully installed boto3-1.35.5 botocore-1.35.5 cohere-5.8.1 couchbase-4.3.0 dataclasses-json-0.6.7 datasets-2.21.0 dill-0.3.8 fastavro-1.9.5 h11-0.14.0 httpcore-1.0.5 httpx-0.27.0 httpx-sse-0.4.0 jmespath-1.0.1 jsonpatch-1.33 jsonpointer-3.0.0 langchain-0.2.14 langchain-community-0.2.12 langchain-experimental-0.0.64 langchain-text-splitters-0.2.2 langchain_cohere-0.2.2 langchain_core-0.2.35 langchain_couchbase-0.1.1 langsmith-0.1.104 marshmallow-3.22.0 multiprocess-0.70.16 mypy-extensions-1.0.0 orjson-3.10.7 parameterized-0.9.0 pyarrow-17.0.0 python-dotenv-1.0.1 s3transfer-0.10.2 tenacity-8.5.0 types-requests-2.32.0.20240712 typing-inspect-0.9.0 xxhash-3.5.0\n"
+          ]
+        }
+      ],
+      "source": [
+        "# Install python-dotenv\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:2271]: Is this accurate?

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 16]
```diff
@@ -0,0 +1,3514 @@
...
+        "!pip install python-dotenv couchbase datasets langchain_core langchain_cohere langchain_couchbase tqdm"
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:2272]: Do you need all these libraries?

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 17]
```diff
@@ -0,0 +1,3514 @@
...
+      "metadata": {
+        "id": "R-SanCZrCLdm"
+      },
+      "execution_count": 3,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# Loading Environment Variables\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:2347]: What is an environment variable?

Comment (nithishr) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb:2347]: https://github.com/couchbase-examples/vector-search-cookbook/pull/1#discussion_r1726671084

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 18]
```diff
@@ -0,0 +1,3514 @@
...
+            "2024-08-26 00:36:06,099 - INFO - Successfully created Cohere LLM with model command\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# Perform Semantic Search\n",
+        "Semantic search in Couchbase involves converting a query and documents into vector representations using a model like OpenAI's embeddings. These vectors capture the semantic meaning of the text and are stored in Couchbase's vector store. When a query is made, Couchbase performs a similarity search by comparing the query vector against the stored document vectors, often using cosine similarity to measure how close they are in meaning. The system retrieves the top k most relevant documents based on these similarity scores.\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:3281]: The similarity metric is user defined

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 19]
```diff
@@ -0,0 +1,3514 @@
...
+    {
+      "cell_type": "code",
+      "source": [
+        "query = \"What caused the 1929 Great Depression?\"\n",
+        "\n",
+        "try:\n",
+        "    start_time = time.time()\n",
+        "    search_results = vector_store.similarity_search_with_score(query, k=10)\n",
+        "    elapsed_time = time.time() - start_time\n",
+        "    results = [{'id': doc.metadata.get('id', 'N/A'), 'text': doc.page_content, 'distance': score}\n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:3298]: Why do you need to do all this processing here especially considering that you are not doing anything with most of the content?

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb [hunk 20]
```diff
@@ -0,0 +1,3514 @@
...
+        "id": "PvuJyXPUFOux",
+        "outputId": "3925085e-2173-42dc-c942-e72857c65832"
+      },
+      "execution_count": 19,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "RAG Response: You appear to have asked a question about the cause of the 1929 Great Depression. I cannot find any direct answer to your question in the provided context. However, it is believed that the stock market crash of 1929, a significant event that triggered the Great Depression, was caused by the excessive production of goods, unequal distribution of wealth, and unregulated market speculation. \n",
```
Comment (nithishr) [inline cohere/RAG_with_Couchbase_and_Cohere.ipynb:3421]: Is this example accurate? The answer is coming entirely from the LLM without any contextual info.

File: cohere/requirements.txt [hunk 21]
```diff
@@ -0,0 +1,7 @@
+couchbase==4.3.0
```
Comment (nithishr) [inline cohere/requirements.txt:1]: Check if you need all the dependencies. You don't need some of them for sure.

**Review by nithishr — CHANGES_REQUESTED** [review]

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 15 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: removed that unnecessary comment.

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 16 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: removed python-dotenv, remaining i do require, we can see that we have used them in the cell below

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 17 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: Changed it to loading sensitive information

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 18 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: understood. changed it

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 19 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: removed this unnecessary processing

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 20 above)
Comment (teetangh) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: Now am using an example that comes from the dataset

File: cohere/requirements.txt (hunk 21 above)
Comment (teetangh) [reply cohere/requirements.txt:1]: I do need all these dependencies.

`couchbase`: I need this package to connect to Couchbase for managing and interacting with the Couchbase cluster, 

`dotenv`: for loading the .env file in local

`tqdm`: for progress display in long waiting loops

`datasets`: contains loadset for TREC dataset

`langchain_core`: I need this package for building and managing the core components of the LangChain framework. This includes setting up and using language models (LLM), handling document processing with Document, defining custom output parsing with StrOutputParser, and creating and managing prompts using ChatPromptTemplate. The RunnablePassthrough is useful for chaining operations together. Additionally, langchain_core includes the ability to manage global settings and caching, which are important for optimizing performance and managing resources across the system.

`langchain_cohere`: contains ChatCohere for llm model , CohereEmbeddings for embeddings generator 

`langchain_couchbase`: contains CouchbaseCache

Comment (nithishr) [reply cohere/requirements.txt:1]: Check what happens when you install `langchain_couchbase` on a fresh environment.

File: cohere/RAG_with_Couchbase_and_Cohere.ipynb (hunk 16 above)
Comment (nithishr) [reply cohere/RAG_with_Couchbase_and_Cohere.ipynb]: You need the libraries but do you need to install all of them? Check out transitive dependencies

File: cohere/requirements.txt (hunk 21 above)
Comment (teetangh) [reply cohere/requirements.txt]: removed couchbase, langchain_core and tqdm

**Review by nithishr — APPROVED** [review]

**Pull request merged**